### PR TITLE
[Distributed] Make new protocol requirements conditionally available

### PR DESCRIPTION
**Pull request by xedin**
- Bring back ad-hoc requirement references in SIL functions to support direct calls to old style ad-hoc requirements
- Choose how to dispatch `decodeNextArgument` based on presence of concrete decoder
- Mark all newly added requirements as available starting from stdlib 5.11

### CODE REVIEW

**Comment by xedin** [comment]
@swift-ci please test

**Comment by ktoso** [comment]
Windows failures

```
Failed Tests (2):
  Swift(windows-x86_64) :: InterfaceHash/edited_method_body.swift
  Swift(windows-x86_64) :: InterfaceHash/edited_property_getter.swift
```

seem unrelated; https://ci-external.swift.org/job/swift-PR-windows/21986/console

**Comment by ktoso** [comment]
@swift-ci please test Windows

File: lib/SIL/IR/SILFunctionBuilder.cpp [hunk 1]
```diff
@@ -275,14 +275,15 @@ void SILFunctionBuilder::addFunctionAttributes(
       F->setDynamicallyReplacedFunction(replacedFunc);
     }
   } else if (constant.isDistributedThunk()) {
-    auto decodeFuncDecl =
+    // It's okay for `decodeFuncDecl` to be null because system could be
+    // generic.
```
Comment (ktoso) [inline lib/SIL/IR/SILFunctionBuilder.cpp:279]: :+1:

**Review by ktoso — APPROVED** [review]

**Comment by ktoso** [comment]
Looks good, I'll also verify with "new compiler, old runtime"

**Pull request merged**